### PR TITLE
[KernelGen][Nvidia] Add rnn_relu operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary

Adds `rnn_relu` support with a Triton implementation for the NVIDIA backend.

The implementation covers the tested single-layer RNN ReLU path and includes forward, direct-wrapper, backward, and large-hidden-size regression coverage. The latest follow-up removed the global `tools/test-op.sh` skip and handles quick CPU mode locally in `tests/test_rnn_relu.py`.

## Review Follow-Up

- Removed `tools/test-op.sh` from the PR diff.
- Cleaned commit history to remove disallowed co-author trailers.
- Added local `cfg.TO_CPU` skip handling in `tests/test_rnn_relu.py` for CUDA-only tests.
- Kept the quick CPU handling operator-scoped instead of changing global test runner behavior.

## Testing

Local validation:

- `python -m pytest -s -x tests/test_rnn_relu.py --ref=cpu --quick`
  - Result: 196 skipped, PASS. This validates quick CPU mode no longer fails while keeping the skip local to this test file.
- `python -m compileall src/flag_gems/ops/rnn_relu.py tests/test_rnn_relu.py`
- `pre-commit run --files benchmark/test_rnn_relu.py conf/operators.yaml src/flag_gems/__init__.py src/flag_gems/ops/__init__.py src/flag_gems/ops/rnn_relu.py src/flag_gems/runtime/backend/_nvidia/tune_configs.yaml tests/test_rnn_relu.py`

Latest GitHub CI on head `cd5a0087`:

| Check | Result |
|---|---|
| `code-style` | SUCCESS |
| `triage` | SUCCESS |
| `preprocess` | SUCCESS |
| `python-op` | SUCCESS |
| `backend-nvidia / test` | SUCCESS |
| Other backend jobs | SKIPPED |
| `license/cla` | SUCCESS |

## Performance

Not benchmarked in the latest GitHub CI rerun. `benchmark/test_rnn_relu.py` is included for core benchmark coverage.

## Multi-Backend Testing

| Backend | Status | Notes |
|---|---|---|
| Nvidia | PASS | `backend-nvidia / test` passed |
| Ascend | Not run | Backend job skipped |
| Enflame | Not run | Backend job skipped |
| Hygon | Not run | Backend job skipped |
| Iluvatar | Not run | Backend job skipped |
| Kunlunxin | Not run | Backend job skipped |
| MetaX | Not run | Backend job skipped |
| MThreads | Not run | Backend job skipped |
| Spacemit | Not run | Backend job skipped |
| Sunrise | Not run | Backend job skipped |
| THead | Not run | Backend job skipped |
| Tsingmicro | Not run | Backend job skipped |

## Files Changed

- `src/flag_gems/ops/rnn_relu.py`: Triton implementation
- `tests/test_rnn_relu.py`: Accuracy, direct-wrapper, backward, large-hidden-size, and local quick CPU skip coverage
- `benchmark/test_rnn_relu.py`: Core benchmark coverage
- `src/flag_gems/runtime/backend/_nvidia/tune_configs.yaml`: NVIDIA tune config
- `src/flag_gems/ops/__init__.py`: Operator export
- `src/flag_gems/__init__.py`: Operator registration
- `conf/operators.yaml`: Operator metadata

### CODE REVIEW

File: src/flag_gems/ops/rnn_relu.py [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+
+
+@libentry()
+@triton.autotune(
+    configs=[
+        triton.Config({"BLOCK_SIZE": 128}, num_stages=4, num_warps=4),
+        triton.Config({"BLOCK_SIZE": 256}, num_stages=4, num_warps=4),
+        triton.Config({"BLOCK_SIZE": 512}, num_stages=4, num_warps=4),
+        triton.Config({"BLOCK_SIZE": 1024}, num_stages=4, num_warps=4),
+    ],
```
Comment (lukatao) [inline src/flag_gems/ops/rnn_relu.py]: move the autotune config into the config file pls

File: src/flag_gems/ops/rnn_relu.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+            dropout,
+            train,
+            bidirectional,
+            batch_first,
+        )
+    else:
+        # Unsupported configuration
+        raise NotImplementedError(
+            "GEMS RNN_RELU only supports single-layer " "unidirectional without dropout"
+        )
```
Comment (lukatao) [inline src/flag_gems/ops/rnn_relu.py]: We need a more general implementation.

**Review by Schopenhauer-loves-Hegel — CHANGES_REQUESTED** [review]

**Review by Schopenhauer-loves-Hegel — CHANGES_REQUESTED** [review]
## FlagGems Automated Review

**Operator:** `rnn_relu`
**Verdict:** REQUEST_CHANGES

### Summary
- Errors: 1
- Warnings: 0
- Suggestions: 0

### Issues (no specific line)
- **[ERROR]** `UNKNOWN`: Commit 中包含 'Co-Authored-By: Claude Opus 4.6 <noreply@anthropic.com>'。FlagGems 不允许在 commit message 中使用 Co-Authored-By — *Fix: 使用 git rebase -i 修改 commit message，删除 Co-Authored-By 行*

File: src/flag_gems/ops/rnn_relu.py (hunk 1 above)
Comment (XDYuanzhuLee) [reply src/flag_gems/ops/rnn_relu.py]: Acknowledged. This is not fixed yet. The autotune config is still inline in the current implementation; it needs to be moved to the repo's config pattern before this PR is ready.

File: src/flag_gems/ops/rnn_relu.py (hunk 2 above)
Comment (XDYuanzhuLee) [reply src/flag_gems/ops/rnn_relu.py]: Agreed. The current implementation is too narrow: it only covers the simple single-layer/unidirectional/no-dropout case. I will treat this as blocked until the implementation is generalized or the PR scope is explicitly changed.

**Comment by XDYuanzhuLee** [comment]
Status update: this PR is blocked rather than fixed. Remaining items are: remove the `Co-Authored-By` trailer from commit history (requires a history rewrite/force-push), move the inline autotune config into the project config pattern, and generalize the implementation beyond the current single-layer/unidirectional/no-dropout case.

**Comment by bin913** [comment]
@XDYuanzhuLee Please fix failing check

File: src/flag_gems/ops/rnn_relu.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+import triton
+import triton.language as tl
+
+from flag_gems.utils import libentry
+
+logger = logging.getLogger(__name__)
+
+
+@libentry()
+@triton.autotune(
```
Comment (bin913) [inline src/flag_gems/ops/rnn_relu.py:32]: **Suggestion**: Large inline autotune config (4 configs hardcoded in kernel)

**Recommended fix:**
Move autotune configs to a config file instead of hardcoding inline.

<sub>Rule: `KERNEL_AUTOTUNE_CONFIG`</sub>

**Review by bin913 — COMMENTED** [review]
## FlagGems Automated Review

**Operator:** `rnn_relu`
**Verdict:** COMMENT

### Summary
- Errors: 0
- Warnings: 1
- Suggestions: 0

**Comment by lukatao** [comment]
fix the unit test error pls

File: src/flag_gems/ops/rnn_relu.py (hunk 3 above)
Comment (XDYuanzhuLee) [reply src/flag_gems/ops/rnn_relu.py:32]: Both review issues are already addressed in the current branch:

1. **Autotune configs**: Already moved from inline to `src/flag_gems/runtime/backend/_nvidia/tune_configs.yaml` (commit 2b098c20). The kernel now loads them via `runtime.get_tuned_config("rnn_relu")`.

2. **Failing check**: The latest commit (6929d5e7, "fix(rnn_relu): align wrapper schema and test marker coverage") addressed test and schema issues.

Validation (current state): `python -m py_compile src/flag_gems/ops/rnn_relu.py` PASS; `pre-commit run --files src/flag_gems/ops/rnn_relu.py` PASS (isort, black, flake8). CI re-run may be needed to confirm the check passes.

**Comment by lukatao** [comment]
resolve confilcts pls

File: conf/operators.yaml [hunk 4]
```diff
@@ -4238,6 +4238,18 @@ ops:
...
+    for:
+      - rnn_relu
+    labels:
+      - aten
+      - KernelGen
+      - NoCPU
+    kind:
+      - Math
+    stages:
+      - alpha: '5.1'
```
Comment (lukatao) [inline conf/operators.yaml]: use 5.3 pls

File: src/flag_gems/ops/rnn_relu.py [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+    weight_ih = weight_ih.contiguous()
+    weight_hh = weight_hh.contiguous()
+    if bias_ih is not None:
+        bias_ih = bias_ih.contiguous()
+    if bias_hh is not None:
+        bias_hh = bias_hh.contiguous()
+
+    grid = (batch_size,)
+
+    with torch.cuda.device(input.device):
```
Comment (lukatao) [inline src/flag_gems/ops/rnn_relu.py]: do not use torch.cuda.device, try flag_gems.device

**Comment by XDYuanzhuLee** [comment]
Fixed. The conflict is resolved. `torch.cuda.device` has been replaced with the proper device handling. The operator stage is set to alpha 5.3. Validation: all CI checks (code-style, triage, preprocess, python-op) are passing.

File: src/flag_gems/ops/rnn_relu.py (hunk 5 above)
Comment (XDYuanzhuLee) [reply src/flag_gems/ops/rnn_relu.py]: Fixed. Replaced `torch.cuda.device` with `flag_gems.device`.

File: conf/operators.yaml (hunk 4 above)
Comment (XDYuanzhuLee) [reply conf/operators.yaml]: Fixed. Changed rnn_relu alpha from 5.4 to 5.3 as requested.

**Comment by bin913** [comment]
@XDYuanzhuLee Please resolve conflicts

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator 贡献审查报告

**Operator**: `rnn_relu`
**PR**: #3558

### 📊 变更统计
- 新增行数: 763
- 删除行数: 3  
- 变更文件: 7

### ✅ 检查清单

| 检查项 | 状态 |
|--------|------|
| 实现文件 (`src/flag_gems/ops/rnn_relu.py`) | ✅ 存在 |
| 测试文件 (`tests/test_rnn_relu.py`) | ✅ 存在 |
| Benchmark 文件 | ✅ 存在 |
| 配置注册 (`operators.yaml`) | ✅ 存在 |
| 模块注册 (`__init__.py`) | ✅ 存在 |
| KernelGen 标注 | ✅ 存在 |
| Triton JIT 装饰器 | ✅ 存在 |
| Metax Backend | ⚠️ 未实现 |

### 📝 审查意见

**代码结构**: 实现文件结构正确

**测试覆盖**: 有测试文件

**性能测试**: 有 benchmark

---

*此审查由自动化脚本根据 FlagGems Operator 贡献规范生成。*
*如有问题请联系维护者进行详细审查。*

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `rnn_relu`  
**PR**: #3558

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 763 |
| Deletions | 3 |
| Files Changed | 7 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/rnn_relu.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ✅ Present: `tests/test_rnn_relu.py` |
| 8 | Test coverage (shapes, dtypes) | ⚠️ May need more test coverage |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_rnn_relu.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ✅ Stability patterns found |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ✅ Docstrings present |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 12 |
| ⚠️ Warning | 9 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Comment by XDYuanzhuLee** [comment]
Thanks. Conflicts resolved by merging origin/master. Validated with `git diff --check`, `py_compile`, and `check_operator --strict` (0 errors, 0 warnings). Commit bda5d5e.

**Review by Dongxu-H — DISMISSED** [review]
## Code Review: RNN ReLU Implementation

I've reviewed the implementation - **algorithm is correct** with some minor cleanup suggestions.

---

### ✅ Algorithm Verification

**RNN ReLU formula**:
```
h_t = relu(x_t @ W_ih^T + h_{t-1} @ W_hh^T + b_ih + b_hh)
```

**Implementation correctness**:

1. **Sequential time step processing** ✅:
```python
for t in range(seq_len):
    # Each step depends on previous hidden state
```

2. **Double-buffering design** ✅:
```python
# Snapshot previous-step hidden state to avoid read-after-write race
for hid_block in range(num_hid_blocks):
    h_val = tl.load(h_scratch_base + h_offs, mask=h_mask, other=0.0)
    tl.store(h_read_base + h_offs, h_val, mask=h_mask)
```

**Why double-buffering?**
- When `hidden_size > BLOCK_SIZE`, one time step needs multiple chunks
- Without double-buffering, later chunks would read **updated values** (current step) instead of **previous step** values
- This ensures correct sequential dependency

3. **Mat-vec computation** ✅:
```python
# Tiled matrix-vector multiplication with CHUNK=64
for inp_start in range(0, input_size, CHUNK):
    w_tile = tl.load(weight_ih_ptr + w_offs, ...)
    x_tile = tl.load(input_ptr + t_inp_off + i_offs, ...)
    ih_acc += tl.sum(w_tile * x_tile[None, :], axis=1)
```

4. **Backward pass** ✅:
```python
# Recompute forward in PyTorch (Triton kernel is opaque to autograd)
with torch.enable_grad():
    h = hx[0].clone()
    outputs = []
    for t_idx in range(seq_len):
        pre_act = torch.addmm(bias_ih, xt, weight_ih.t())
        pre_act += torch.addmm(bias_hh, h, weight_hh.t())
        h = torch.relu(pre_act)
        outputs.append(h)
    
    all_weight_grads = torch.autograd.grad(...)
```

---

### ⚠️ Minor Issues

#### 1. operators.yaml Contains Unrelated Changes

```yaml
# These operators appear in the diff but shouldn't be part of this PR
+  - id: normal_
+  - id: renorm
+  - id: renorm_
+  - id: reflection_pad1d_backward
+  - id: resize_output
+  - id: rnn_relu  # ← Target change
+  - id: rot90
+  - id: smooth_l1_loss
```

**Suggestion**: Verify if these are due to file re-sorting or merge conflict. Consider re-merging master to clean up.

---

#### 2. Test Log Files Should Not Be in PR

```
result-s_benchmark_test_linalg_slogdet--level_core--record_log-k_test_linalg_slogdet--tb_long.log
result-s_benchmark_test_linalg_slogdet--level_core--record_log.log
result-x-s_benchmark_test_linalg_slogdet--level_core--record_log.log
```

**Suggestion**: Remove these benchmark log files from the PR. They are unrelated to `rnn_relu`.

---

#### 3. Missing Performance Data

PR description states:
> Performance: Not benchmarked in the latest GitHub CI rerun.

**Suggestion**: Run benchmark and report performance results.

---

#### 4. Single-layer Unidirectional Limitation

```python
if num_layers > 1 or bidirectional:
    raise NotImplementedError(
        "GEMS RNN_RELU only supports single-layer unidirectional"
    )
```

**Suggestion**: Document this limitation clearly in the PR description and consider future work to support multi-layer/bidirectional.

---

### ✅ What's Good

1. **Algorithm correct** ✅ - Sequential RNN ReLU correctly implemented
2. **Double-buffering** ✅ - Handles `hidden_size > BLOCK_SIZE` correctly
3. **Backward correct** ✅ - PyTorch recompute + autograd
4. **Autotune config** ✅ - BLOCK_SIZE: 128, 256, 512, 1024
5. **Large hidden_size regression test** ✅ - Tests hidden_size=128, 256, 512
6. **CI passes** ✅ - python-op, code-style, backend-nvidia all PASS
7. **Complete registration** ✅ - operators.yaml, __init__.py, ops/__init__.py
8. **KernelGen attribution** ✅
9. **Conflicts resolved** ✅

---

### 📊 Summary

| Aspect | Status |
|--------|--------|
| Algorithm correctness | ✅ Correct |
| Double-buffering design | ✅ Correct |
| Backward pass | ✅ Correct |
| operators.yaml cleanup | ⚠️ Contains unrelated changes |
| Test log files | ⚠️ Should be removed |
| Performance data | ⚠️ Missing |
| Single-layer limitation | ⚠️ Document clearly |
| CI status | ✅ All PASS |
| Registration | ✅ Complete |

---

### 🎯 Conclusion

**Approve** ✅ with minor cleanup suggestions.

The algorithm is correct and the double-buffering design properly handles the sequential dependency. The issues mentioned are non-blocking cleanup items.

Non-blocking suggestions:
1. Clean up operators.yaml (remove unrelated changes)
2. Remove test log files
3. Run benchmark and report performance
4. Document single-layer unidirectional limitation

Good work on the implementation!

**Review by Schopenhauer-loves-Hegel — DISMISSED** [review]

File: tests/test_rnn_relu.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        xt = inp_r[t_idx, :, :]
+        pre_act = torch.addmm(bias_ih, xt, weight_ih.t())
+        pre_act += torch.addmm(bias_hh, h, weight_hh.t())
+        h = torch.relu(pre_act)
+        outputs.append(h)
+    out_r = torch.stack(outputs, dim=0)
+    hid_r = h.unsqueeze(0)
+    (out_r.sum() + hid_r.sum()).backward()
+
+    torch.testing.assert_close(inp_g.grad, inp_r.grad)
```
Comment (Caeruleann) [inline tests/test_rnn_relu.py]: Why not use `utils.gems_assert_close`

Comment (XDYuanzhuLee) [reply tests/test_rnn_relu.py]: Done. The specific line you flagged (now line 164) was already using `utils.gems_assert_close` after a prior update, and I've also replaced the remaining `torch.testing.assert_close` calls in `test_rnn_relu_large_hidden` with `utils.gems_assert_close`. All assertions in the test file now use the project's utility.

**Review by Caeruleann — APPROVED** [review]

**Review by Dongxu-H — COMMENTED** [review]
**Redundant constant definition** (lines 28 and 69 in rnn_relu.py)
   - Line 30: `_CHUNK = 64` (module-level)
   - Line 76: `CHUNK: tl.constexpr = 64` (kernel-level)
   
   The outer `_CHUNK` is never used in the kernel (kernel uses inner `CHUNK`). Either use the outer constant or remove the duplicate for clarity.
